packages/react-router/modules/Route.js

```diff
@@ -62,15 +63,15 @@ class Route extends React.Component {
     )
 
     this.setState({
-      match: this.computeMatch(nextProps)
+      match: this.computeMatch(nextProps, nextContext)
     })
   }
 
   render() {
     const { match } = this.state
     const { children, component, render } = this.props
-    const { history } = this.context
-    const { location } = history
+    const { history, route } = this.context
+    const location = this.props.location || route.location
```

Comment:
Just making sure that the component rendered by the route receives the correct location.